### PR TITLE
Fix `eth_getBlockTransactionCountByNumber` crashed on block not found

### PR DESCRIPTION
**Pull request by calvinaco**
Closes: #253

## Description

Add `nil` block handling to related RPC methdos

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: ethereum/rpc/namespaces/eth/api.go [hunk 1]
```diff
@@ -659,6 +667,11 @@ func (e *PublicAPI) GetTransactionByBlockHashAndIndex(hash common.Hash, idx hexu
 		return nil, nil
 	}
 
+	if resBlock.Block == nil {
+		e.logger.WithError(err).Debugln("block not found", "hash", hash.Hex())
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:671]:
```suggestion
		e.logger.Debugln("block not found", "hash", hash.Hex())
```

File: ethereum/rpc/namespaces/eth/api.go [hunk 2]
```diff
@@ -286,6 +290,10 @@ func (e *PublicAPI) GetBlockTransactionCountByNumber(blockNum rpctypes.BlockNumb
 		return nil
 	}
 
+	if resBlock.Block == nil {
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:293]: can you add a log here too?

File: ethereum/rpc/namespaces/eth/api.go [hunk 3]
```diff
@@ -274,6 +274,10 @@ func (e *PublicAPI) GetBlockTransactionCountByHash(hash common.Hash) *hexutil.Ui
...
+		return nil
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:278]: ditto, log

File: ethereum/rpc/namespaces/eth/api.go [hunk 4]
```diff
@@ -710,6 +723,11 @@ func (e *PublicAPI) GetTransactionByBlockNumberAndIndex(blockNum rpctypes.BlockN
 		return nil, nil
 	}
 
+	if resBlock.Block == nil {
+		e.logger.WithError(err).Debugln("block not found", "height", blockNum.Int64())
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:727]:
```suggestion
		e.logger.Debugln("block not found", "height", blockNum.Int64())
```

File: ethereum/rpc/namespaces/eth/api.go (hunk 1 above)
Comment (calvinaco) [reply ethereum/rpc/namespaces/eth/api.go:671]: Thanks!

File: ethereum/rpc/namespaces/eth/api.go (hunk 4 above)
Comment (calvinaco) [reply ethereum/rpc/namespaces/eth/api.go:727]: Thanks!

File: ethereum/rpc/namespaces/eth/api.go (hunk 3 above)
Comment (calvinaco) [reply ethereum/rpc/namespaces/eth/api.go:280]: Log added, thanks!

File: ethereum/rpc/namespaces/eth/api.go (hunk 2 above)
Comment (calvinaco) [reply ethereum/rpc/namespaces/eth/api.go:296]: Log added, thanks!

File: ethereum/rpc/namespaces/eth/api.go [hunk 5]
```diff
@@ -280,9 +286,15 @@ func (e *PublicAPI) GetBlockTransactionCountByHash(hash common.Hash) *hexutil.Ui
 
 // GetBlockTransactionCountByNumber returns the number of transactions in the block identified by number.
 func (e *PublicAPI) GetBlockTransactionCountByNumber(blockNum rpctypes.BlockNumber) *hexutil.Uint {
-	e.logger.Debugln("eth_getBlockTransactionCountByNumber", "block number", blockNum)
+	e.logger.Debugln("eth_getBlockTransactionCountByNumber", "block number", blockNum.Int64())
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:289]:
```suggestion
	e.logger.Debugln("eth_getBlockTransactionCountByNumber", "height", blockNum.Int64())
```

File: ethereum/rpc/namespaces/eth/api.go [hunk 6]
```diff
@@ -280,9 +286,15 @@ func (e *PublicAPI) GetBlockTransactionCountByHash(hash common.Hash) *hexutil.Ui
...
 	resBlock, err := e.clientCtx.Client.Block(e.ctx, blockNum.TmHeight())
 	if err != nil {
+		e.logger.WithError(err).Debugln("block not found", "block number", blockNum.Int64())
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:292]:
```suggestion
		e.logger.WithError(err).Debugln("block not found", "height", blockNum.Int64())
```

File: ethereum/rpc/namespaces/eth/api.go [hunk 7]
```diff
@@ -280,9 +286,15 @@ func (e *PublicAPI) GetBlockTransactionCountByHash(hash common.Hash) *hexutil.Ui
...
+		return nil
+	}
+
+	if resBlock.Block == nil {
+		e.logger.Debugln("block not found", "block number", blockNum.Int64())
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:297]:
```suggestion
		e.logger.Debugln("block not found", "height", blockNum.Int64())
```

**Review by fedekunze — APPROVED** [review]

**Comment by fedekunze** [comment]
@calvinaco can you fix the conflicts 🙏

File: ethereum/rpc/namespaces/eth/api.go [hunk 8]
```diff
@@ -282,9 +288,15 @@ func (e *PublicAPI) GetBlockTransactionCountByHash(hash common.Hash) *hexutil.Ui
 
 // GetBlockTransactionCountByNumber returns the number of transactions in the block identified by number.
 func (e *PublicAPI) GetBlockTransactionCountByNumber(blockNum rpctypes.BlockNumber) *hexutil.Uint {
-	e.logger.Debug("eth_getBlockTransactionCountByNumber", "block number", blockNum)
+  e.logger.Debug("eth_getBlockTransactionCountByNumber", "block number", blockNum.Int64())
 	resBlock, err := e.clientCtx.Client.Block(e.ctx, blockNum.TmHeight())
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:292]:
```suggestion
	e.logger.Debug("eth_getBlockTransactionCountByNumber", "block number", blockNum.Int64())
	resBlock, err := e.clientCtx.Client.Block(e.ctx, blockNum.TmHeight())
```

File: ethereum/rpc/namespaces/eth/api.go [hunk 9]
```diff
@@ -282,9 +288,15 @@ func (e *PublicAPI) GetBlockTransactionCountByHash(hash common.Hash) *hexutil.Ui
...
-	e.logger.Debug("eth_getBlockTransactionCountByNumber", "block number", blockNum)
+	e.logger.Debug("eth_getBlockTransactionCountByNumber", "block number", blockNum.Int64())
 	resBlock, err := e.clientCtx.Client.Block(e.ctx, blockNum.TmHeight())
 	if err != nil {
+		e.logger.WithError(err).Debugln("block not found", "height", blockNum.Int64())
+		return nil
+	}
+
+	if resBlock.Block == nil {
+		e.logger.Debugln("block not found", "height", blockNum.Int64())
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:299]:
```suggestion
		e.logger.Debug("block not found", "height", blockNum.Int64(), "error", err.Error())
		return nil
	}

	if resBlock.Block == nil {
		e.logger.Debug("block not found", "height", blockNum.Int64())
```

File: ethereum/rpc/namespaces/eth/api.go [hunk 10]
```diff
@@ -668,6 +680,11 @@ func (e *PublicAPI) GetTransactionByBlockHashAndIndex(hash common.Hash, idx hexu
 		return nil, nil
 	}
 
+	if resBlock.Block == nil {
+		e.logger.Debugln("block not found", "hash", hash.Hex())
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:684]:
```suggestion
		e.logger.Debug("block not found", "hash", hash.Hex())
```

File: ethereum/rpc/namespaces/eth/api.go [hunk 11]
```diff
@@ -706,6 +723,11 @@ func (e *PublicAPI) GetTransactionByBlockNumberAndIndex(blockNum rpctypes.BlockN
 		return nil, nil
 	}
 
+	if resBlock.Block == nil {
+		e.logger.Debugln("block not found", "height", blockNum.Int64())
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:727]:
```suggestion
		e.logger.Debug("block not found", "height", blockNum.Int64())
```

File: ethereum/rpc/namespaces/eth/api.go [hunk 12]
```diff
@@ -273,6 +273,12 @@ func (e *PublicAPI) GetBlockTransactionCountByHash(hash common.Hash) *hexutil.Ui
 
 	resBlock, err := e.clientCtx.Client.BlockByHash(e.ctx, hash.Bytes())
 	if err != nil {
+		e.logger.WithError(err).Debugln("block not found", "hash", hash.Hex())
+		return nil
+	}
+
+	if resBlock.Block == nil {
+		e.logger.Debugln("block not found", "hash", hash.Hex())
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:281]:
```suggestion
		e.logger.Debug("block not found", "hash", hash.Hex(), "error", err.Error())
		return nil
	}

	if resBlock.Block == nil {
		e.logger.Debug("block not found", "hash", hash.Hex())
```

File: ethereum/rpc/namespaces/eth/api.go [hunk 13]
```diff
@@ -282,9 +288,15 @@ func (e *PublicAPI) GetBlockTransactionCountByHash(hash common.Hash) *hexutil.Ui
 
 // GetBlockTransactionCountByNumber returns the number of transactions in the block identified by number.
 func (e *PublicAPI) GetBlockTransactionCountByNumber(blockNum rpctypes.BlockNumber) *hexutil.Uint {
-	e.logger.Debug("eth_getBlockTransactionCountByNumber", "block number", blockNum)
+	e.logger.Debug("eth_getBlockTransactionCountByNumber", "block number", blockNum.Int64())
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:291]:
```suggestion
	e.logger.Debug("eth_getBlockTransactionCountByNumber", "height", blockNum.Int64())
```

**Pull request merged**